src/main/java/randoop/generation/OrienteeringSelection.java

```diff
@@ -0,0 +1,119 @@
+
+  /**
+   * Cache for the square root of the number of method calls in a sequence of statements. Once
+   * computed for a given sequence, the value is never updated.
+   */
+  private final Map<Sequence, Double> sequenceMethodSizeSqrt = new HashMap<>();
+
+  /**
+   * Map from a sequence to its approximate execution time in milliseconds. Once computed for a
+   * given sequence, the value is never updated.
+   */
+  private final Map<Sequence, Long> sequenceExecutionTime = new HashMap<>();
+
+  /**
+   * The sequence that was last selected by Orienteering. A reference is kept so that we can update
+   * its execution time once {@link ForwardGenerator} finishes executing the newly constructed
+   * sequence.
+   */
+  private Sequence lastSelectedSequence;
+
+  /**
+   * Bias input selection towards lower cost sequences.
+   *
+   * @param candidates sequences to choose from
+   * @return the chosen sequence
+   */
+  @Override
+  public Sequence selectInputSequence(SimpleList<Sequence> candidates) {
+    updateWeightMapForCandidates(candidates);
+    lastSelectedSequence = Randomness.randomMemberWeighted(candidates, weightMap);
+    incrementCountInMap(sequenceSelectionCount, lastSelectedSequence);
+    return lastSelectedSequence;
+  }
+
+  /**
+   * Update the weights of the candidates in the given list
+   *
+   * @param candidates list of candidates
+   */
+  private void updateWeightMapForCandidates(SimpleList<Sequence> candidates) {
+    for (int i = 0; i < candidates.size(); i++) {
+      Sequence candidate = candidates.get(i);
+
+      // If we haven't computed the square root of the method size of this sequence,
+      // compute it and permanently store it.
+      Double methodSizeSqrt = sequenceMethodSizeSqrt.get(candidate);
+      if (methodSizeSqrt == null) {
+        methodSizeSqrt = Math.sqrt(candidate.numMethodCalls());
+        sequenceMethodSizeSqrt.put(candidate, methodSizeSqrt);
+      }
+
+      // If this sequence has not been selected yet, don't compute its weight and
+      // use its default value.
+      Integer selectionCount = sequenceSelectionCount.get(candidate);
+      if (candidate == null) {
+        continue;
+      }
+
+      // If this sequence has not been executed yet, don't compute its weight and
+      // use its default value.
+      Long executionTime = sequenceExecutionTime.get(candidate);
+      if (executionTime == null) {
+        continue;
+      }
+
+      // Recompute and update this sequence's weight.
+      weightMap.put(candidate, 1.0 / (selectionCount * executionTime * methodSizeSqrt));
+    }
+  }
+
+  /**
+   * Assign the last selected sequence's execution time to be that of this recently executed
+   * executable sequence.
+   *
+   * @param eSeq the recently executed sequence
+   */
+  @Override
+  public void assignExecTimeForSequence(ExecutableSequence eSeq) {
+    Long executionTime = sequenceExecutionTime.get(lastSelectedSequence);
```

Comment:
What is the relationship between `eSeq` and `lastSelectedSequence`?
Are they guaranteed to be the same?  If so, I would assert that, or else eliminate the local variable or the parameter.  (And in that case maybe the map isn't necessary -- the `exectime` field might be sufficient.)
If they are not the same, then document their relationship.

Reply:
Last selected sequence is expected to be a subsequence of eSeq. This comment helped me realize that there can be multiple input sequences used in constructing a new executable sequence. ForwardGenerator now maintains a set of sequences used as input in constructing a new sequence and passes it to Orienteering with the exec time of the executed eSeq.